ratis-server/src/main/java/org/apache/ratis/server/leader/LogAppenderBase.java

```diff
@@ -271,18 +271,21 @@ protected ReferenceCountedObject<AppendEntriesRequestProto> nextAppendEntriesReq
       return null;
     }
 
-    final List<LogEntryProto> protos = buffer.pollList(getHeartbeatWaitTimeMs(), EntryWithData::getEntry,
-        (entry, time, exception) -> LOG.warn("Failed to get " + entry
-            + " in " + time.toString(TimeUnit.MILLISECONDS, 3), exception));
-    for (EntryWithData entry : buffer) {
-      // Release remaining entries.
-      offered.remove(entry.getIndex()).release();
+    try {
```

Comment:
Let's `try` only `pollList`, i.e. 
```java
    final List<LogEntryProto> protos;
    try {
      protos = buffer.pollList(getHeartbeatWaitTimeMs(), EntryWithData::getEntry,
          (entry, time, exception) -> LOG.warn("Failed to get {} in {}",
              entry, time.toString(TimeUnit.MILLISECONDS, 3), exception));
    } finally {
      for (EntryWithData entry : buffer) {
        // Release remaining entries.
        offered.remove(entry.getIndex()).release();
      }
      buffer.clear();
    }
    assertProtos(protos, followerNext, previous, snapshotIndex);
    AppendEntriesRequestProto appendEntriesProto =
        leaderState.newAppendEntriesRequestProto(follower, protos, previous, callId);
    return ReferenceCountedObject.delegateFrom(offered.values(), appendEntriesProto);
  }
```

Reply:
Yes, this is proper way, or "appendEntriesProto" will delegate on more records in `offered` since the leftovers in buffer not removed from `offered` yet.

Reply:
Good idea, thank Nicholas.